sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -860,27 +865,93 @@ public TableDestination apply(ValueInSingleWindow<T> value) {
     }
 
     /**
+     * A function that returns the same schema for every table.
+     */
+    static class ConstantSchemaFunction extends SchemaFunction {
+      private final @Nullable ValueProvider<String> jsonSchema;
+
+      private  ConstantSchemaFunction(@Nullable ValueProvider<String> jsonSchema) {
+        this.jsonSchema = jsonSchema;
+      }
+
+      public static ConstantSchemaFunction fromTableSchema(ValueProvider<TableSchema> tableSchema) {
+        if (tableSchema.isAccessible()) {
+          // TableSchema isn't Serializable, so convert to JSON.
+          return new ConstantSchemaFunction(StaticValueProvider.of(
+              BigQueryHelpers.toJsonString(tableSchema.get())));
+        } else {
+          return new ConstantSchemaFunction(
+              NestedValueProvider.of(tableSchema, new TableSchemaToJsonSchema()));
+        }
+      }
+
+      public static ConstantSchemaFunction fromJsonSchema(ValueProvider<String> jsonSchema) {
+        return new ConstantSchemaFunction(jsonSchema);
+      }
+
+      @Override
+      public TableSchema apply(TableDestination table) {
+        if (jsonSchema == null) {
+          return  null;
+        } else {
+          return BigQueryHelpers.fromJsonString(jsonSchema.get(), TableSchema.class);
+        }
+      }
+    }
+
+    /**
      * Uses the specified schema for rows to be written.
      *
      * <p>The schema is <i>required</i> only if writing to a table that does not already
      * exist, and {@link CreateDisposition} is set to
      * {@link CreateDisposition#CREATE_IF_NEEDED}.
      */
     public Write<T> withSchema(TableSchema schema) {
-      return toBuilder()
-          .setJsonSchema(StaticValueProvider.of(BigQueryHelpers.toJsonString(schema)))
-          .build();
+      return withSchemaFunction(ConstantSchemaFunction.fromTableSchema(
+          StaticValueProvider.of(schema)));
+    }
+
+    public Write<T> withSchema(ValueProvider<TableSchema> schema) {
```

Comment:
These other methods should also have javadoc (not necessarily copying the one above, but referring to it at least)